keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md

```diff
@@ -0,0 +1,834 @@
+Namespace deletion specifically uses `metav1.DeletePropagationBackground` and all resources would be deleted and the ownerReference
+dependencies would be handled by the garbage collection.
+
+In Kubernetes, `ownerReferences` define a parent-child relationship where child resources are automatically deleted when the parent is removed.
+This is mostly handled by garbage collection. While namespace deletion, the `ownerReferences` is not part of the consideration and
+`NamespaceDeletionOrder` group will be honored while deleting resources as what it is currently. The garbage collector controller will make sure  
+no child resources still existing after the parent resource deleted.
+
+
+### Risks and Mitigations
+
+#### Dependency cycle
+
+The introduction of deletion order could potentially cause dependency loops especially when finalizers are 
+specified against deletion priority.
+
+When a lack of progress detected(maybe caused by the dependency cycle described above), the options would be:
+- Fall back to the previous behavior.
+  - Pros: The deletion would not be blocked; no breaking changes;
+  - Cons: Security concern remains unaddressed or could be bypassed
+  
+- Return error after retry.
+  - Pros: Make sure the security concern being addressed by always honor the deletion order
+  - Cons: Block namespace deletion if dependency cycle exists.  
+
+Mitigation: A proper fallback mechanism would be introduced to make sure the namespace deletion process would not be 
+hanging forever because of potential dependency cycle.
+
+Refer to the section `Handling Cyclic Dependencies` for more details.
+
+#### Instance from same resources want different deletion order
+
+The current proposal is to have namespace deletion order per resource. It is possible that the instances 
+with same resources want different namespace deletion order. 
+
+The existing mechanism is to use random order while deletion, the proposal does not make things worse. 
+We could possibly introduce a way to let individual instance be able to specify the deletion order later if certain requests is commonly asked.
+
+## Design Details
+
+### DeletionOrderPriority Mechanism
+
+For the namespace deletion process, we would like to have the resources associated with this namespace be deleted in order of:
+- Workload controllers
+- Workload
+- Default
+- Policies
+
+Each resource type will be assigned a  `NamespaceDeletionOrder` value(<TOBESOLVED - @cici37>int value or string value). To define the DeletionOrder, the options are:
+
+1. Add a field into APIResource (not mutable) so it could be observed easily(pkg/registry):
+
+```
+type APIResource struct {
+……
+`NamespaceDeletionOrder int64 `json:"namespaceDeletionOrder" protobuf:"varint,11,opt,name=namespaceDeletionOrder"``
+}
+```
+
+In this case, the `NamespaceDeletionOrder` will be associated with the resources naturally and the resource which does not have an opinion will default to a value.
+
+- Pros:
+  - Having it configurable
```

Comment:
I think of this as can also be a  Con, since exposing the capacity to establish random deletion priorities between different clusters will make the namespace deletion different per cluster

Reply:
I don't think anyone wants it to be user-configurable.  The question is whether the type defines it's own classification, or whether we define that external to the type.  I think it is intrinsic to the type, but am not going to die on this hill today.  I feel like types SHOULD BE tightly sealed plugins, and any time we store information about those types outside of the type itself, it's a smell.

Reply:
> I don't think anyone wants it to be user-configurable. The question is whether the type defines it's own classification, or whether we define that external to the type. I think it is intrinsic to the type, but am not going to die on this hill today. I feel like types SHOULD BE tightly sealed plugins, and any time we store information about those types outside of the type itself, it's a smell.

When you say type, only core types, not CRDs?

Reply:
For now core, but when CRDs get this, they would be the same.  CRDs which implement policies could denote that.

Reply:
That's at least one bit of user config? :-)

Point taken though, we should probably at least spell out how we're leaving room to add that later without resolving details such as "what is the CRD opt-in field" yet?